cpp/src/arrow/engine/substrait/extension_types.cc

```diff
@@ -0,0 +1,307 @@
+  return utf8();
+}
+static auto kVarCharExtensionParamsProperties = internal::MakeProperties(
+    internal::DataMember("length", &VarCharExtensionParams::length));
+
+using VarCharType =
+    SimpleExtensionType<kVarCharExtensionName, VarCharExtensionParams,
+                        decltype(kVarCharExtensionParamsProperties),
+                        kVarCharExtensionParamsProperties, VarCharGetStorage>;
+
+constexpr util::string_view kIntervalYearExtensionName = "interval_year";
+struct IntervalYearExtensionParams {};
+std::shared_ptr<DataType> IntervalYearGetStorage(const IntervalYearExtensionParams&) {
+  return fixed_size_list(int32(), 2);
+}
+static auto kIntervalYearExtensionParamsProperties = internal::MakeProperties();
+
+using IntervalYearType =
+    SimpleExtensionType<kIntervalYearExtensionName, IntervalYearExtensionParams,
+                        decltype(kIntervalYearExtensionParamsProperties),
+                        kIntervalYearExtensionParamsProperties, IntervalYearGetStorage>;
+
+constexpr util::string_view kIntervalDayExtensionName = "interval_day";
+struct IntervalDayExtensionParams {};
+std::shared_ptr<DataType> IntervalDayGetStorage(const IntervalDayExtensionParams&) {
+  return fixed_size_list(int32(), 2);
+}
+static auto kIntervalDayExtensionParamsProperties = internal::MakeProperties();
+
+using IntervalDayType =
+    SimpleExtensionType<kIntervalDayExtensionName, IntervalDayExtensionParams,
+                        decltype(kIntervalDayExtensionParamsProperties),
+                        kIntervalDayExtensionParamsProperties, IntervalDayGetStorage>;
+
+}  // namespace
+
+std::shared_ptr<DataType> uuid() { return UuidType::Make({}); }
+
+std::shared_ptr<DataType> fixed_char(int32_t length) {
+  return FixedCharType::Make({length});
+}
+
+std::shared_ptr<DataType> varchar(int32_t length) { return VarCharType::Make({length}); }
+
+std::shared_ptr<DataType> interval_year() { return IntervalYearType::Make({}); }
+
+std::shared_ptr<DataType> interval_day() { return IntervalDayType::Make({}); }
+
+bool UnwrapUuid(const DataType& t) {
+  if (auto params = UuidType::GetIf(t)) {
+    return true;
+  }
+  return false;
+}
+
+util::optional<int32_t> UnwrapFixedChar(const DataType& t) {
+  if (auto params = FixedCharType::GetIf(t)) {
+    return params->length;
+  }
+  return util::nullopt;
+}
+
+util::optional<int32_t> UnwrapVarChar(const DataType& t) {
+  if (auto params = VarCharType::GetIf(t)) {
+    return params->length;
+  }
+  return util::nullopt;
+}
+
+bool UnwrapIntervalYear(const DataType& t) {
+  if (auto params = IntervalYearType::GetIf(t)) {
+    return true;
+  }
+  return false;
+}
+
+bool UnwrapIntervalDay(const DataType& t) {
+  if (auto params = IntervalDayType::GetIf(t)) {
+    return true;
+  }
+  return false;
+}
+
+struct ExtensionSet::Impl {
+  struct IdHash {
+    size_t operator()(Id id) const {
+      internal::StringViewHash hash;
+      auto out = hash(id.uri);
```

Comment:
I had to annotate with `size_t out` locally otherwise:

```
/Users/kszucs/Workspace/arrow/cpp/src/arrow/engine/substrait/extension_types.cc:151:7: error: no matching function for call to 'hash_combine'
      internal::hash_combine(out, hash(id.name));
      ^~~~~~~~~~~~~~~~~~~~~~
/Users/kszucs/Workspace/arrow/cpp/src/arrow/util/hash_util.h:57:13: note: candidate function template not viable: no known conversion from 'unsigned long long' to 'std::size_t &' (aka 'unsigned long &') for 1st argument
inline void hash_combine(std::size_t& seed, T const& v) {
            ^
```